packages/esm-patient-chart-app/src/visit/past-visit-overview.component.tsx

```diff
@@ -0,0 +1,194 @@
+import { Visit } from '@openmrs/esm-api';
+import { createErrorHandler } from '@openmrs/esm-error-handling';
+import React, { useState, useEffect, useMemo, useReducer, useCallback } from 'react';
+import DataTable, {
+  TableContainer,
+  Table,
+  TableBody,
+  TableHead,
+  TableRow,
+  TableHeader,
+  TableCell,
+  DataTableHeader,
+  DataTableRow,
+} from 'carbon-components-react/es/components//DataTable';
+import DataTableSkeleton from 'carbon-components-react/es/components/DataTableSkeleton';
+import OverflowMenu from 'carbon-components-react/es/components/OverflowMenu';
+import OverflowMenuItem from 'carbon-components-react/es/components/OverflowMenuItem';
+import { useTranslation } from 'react-i18next';
+import dayjs from 'dayjs';
+import { ErrorState } from '@openmrs/esm-patient-common-lib';
+import { attach, detach, getStartedVisit, getVisitsForPatient, VisitMode, VisitStatus } from '@openmrs/esm-framework';
+import styles from './past-visit-overview.component.scss';
+import Button from 'carbon-components-react/es/components/Button';
+
+// TO DO Abstract this logic for state machines to a hook
+enum ActionTypes {
+  pending = 'pending',
+  resolved = 'resolved',
+  error = 'error',
+}
+
+interface ViewState {
+  status: string;
+}
+
+interface Pending {
+  type: ActionTypes.pending;
+}
+
+interface Error {
+  type: ActionTypes.error;
+}
+
+interface Resolved {
+  type: ActionTypes.resolved;
+}
+
+type Action = Pending | Error | Resolved;
+
+function viewStateReducer(state: ViewState, action: Action): ViewState {
+  switch (action.type) {
+    case ActionTypes.pending:
+      return {
+        status: 'pending',
+      };
+    case ActionTypes.resolved:
+      return { status: 'resolved' };
+    case ActionTypes.error:
+      return { status: 'error' };
+    default:
+      return state;
+  }
+}
```

Comment:
Note that this is basically an identity function, `x => x`. That's a good hint that we can simplify.

There are two routes we could take this:

1. We could replace the `useReducer` line on line 71 with `const [viewState, setViewState] = useState<ViewState>('pending');`, get rid of the `viewStateReducer` function and all the ActionTypes, and replace `dispatch` calls with `setViewState`.

OR

2. We could go all-in on the reducer pattern. The reducer would manage the status *as well as* the data and error states. You'd delete lines 72 and 73 and make this reducer something like

```suggestion
function reducer(state: PastVisitOverviewState, action: Action): PastVisitOverviewState {
  switch (action.type) {
    case ActionTypes.pending:
      return {
        status: 'pending',
        ...state
      };
    case ActionTypes.resolved:
      return {
        status: 'resolved',
        patientPastVisits: action.payload,
        error: null
      };
    case ActionTypes.error:
      return {
        status: 'error',
        patientPastVisits: null,
        error: action.payload
      };
    default:
      return state;
  }
}
```

and then

```typescript
const [{ status, patientPastVisits, error }, dispatch ] = useReducer(reducer, { status: 'pending' });
```

Hopefully that's clear!